File: Classes/server/TKBuzzInfoProvider.m
//
//  BHBuzzInfoProvider.m
//  TripPlanner
//
//  Created by Adrian Schoenig on 29/11/12.
//
//

#import "TKBuzzInfoProvider.h"

#import <TripKit/TKTripKit.h>
#import <TripKit/TripKit-Swift.h>

@implementation TKBuzzInfoProvider


- (void)downloadContentOfService:(Service *)service
							forEmbarkationDate:(NSDate *)date
												inRegion:(SVKRegion *)regionOrNil
											completion:(SGServiceCompletionBlock)completion
{
  NSParameterAssert(service);
  NSParameterAssert(date);
  NSParameterAssert(completion);
  
  ZAssert(service.managedObjectContext.parentContext != nil || [NSThread isMainThread], @"Not on the right thread!");
  

  ZAssert(service.managedObjectContext, @"Service with a context needed.");
  
  if (service.isRequestingServiceData) {
    return; // don't send multiple requests
  }
  
  service.isRequestingServiceData = YES;
	SVKServer *server = [SVKServer sharedInstance];
  [server requireRegions:
   ^(NSError *error) {
     if (error) {
       DLog(@"Error fetching regions: %@", error);
       completion(service, NO);
       return;
     }
     
     SVKRegion *region = regionOrNil ?: service.region;
     if (! region) {
       completion(service, NO);
       return;
     }
     
     NSString *operatorName = service.operatorName ?: @"";
     
     // construct the parameters
     ZAssert(service && service.managedObjectContext, @"Service with a context needed.");
     NSDictionary *paras = @{
                             @"region"						: region.name,
                             @"serviceTripID"	    : service.code,
                             @"operator"	        : operatorName,
                             @"embarkationDate"	  : @([date timeIntervalSince1970]),
                             @"encode"						: @(YES)
                             };
     
     // now send it off to the server
     __weak typeof(self) weakSelf = self;
     [server hitSkedGoWithMethod:@"GET"
                            path:@"service.json"
                      parameters:paras
                          region:region
                  callbackOnMain:NO
                         success:
      ^(NSInteger status, id responseObject, NSData *data) {
#pragma unused(status, data)
        typeof(self) strongSelf = weakSelf;
        if (!strongSelf) {
          return;
        }
        
        [service.managedObjectContext performBlock:^{
          service.isRequestingServiceData = NO;
          if (! responseObject[@"error"]) {
            ZAssert(service && service.managedObjectContext, @"Service with a context needed.");
            [strongSelf addContentToService:service
                               fromResponse:responseObject];
            completion(service, YES);
            
          } else {
            completion(service, NO);
          }
        }];
      }
                         failure:
      ^(NSError *operationError) {
#pragma unused(operationError)
        DLog(@"Error response: %@", operationError);
        [service.managedObjectContext performBlock:^{
          service.isRequestingServiceData = NO;
          completion(service, NO);
        }];
      }];
   }];
}

+ (NSError *)errorForUserForBrokenStop
{
  NSDictionary *info = @{
                         NSLocalizedDescriptionKey: NSLocalizedStringFromTableInBundle(@"Could not find transit stop.", @"TripKit", [TKTripKit bundle], "Error title when server could not find a given transit stop."),
                         NSLocalizedRecoverySuggestionErrorKey: NSLocalizedStringFromTableInBundle(@"Search for this stop again or try again later..", @"TripKit", [TKTripKit bundle], "Error recovery suggestion for when when server could not find a given transit stop."),
                         };
  return [NSError errorWithDomain:@"com.buzzhives.TripKit" code:831571 userInfo:info];
}

+ (void)fillInStop:(StopLocation *)stop
             named:(nullable NSString *)name
        completion:(void (^)(NSError *))completion
{
  NSParameterAssert(stop);
  NSParameterAssert(completion);
  
  // now send it off to the server
  SVKServer *server = [SVKServer sharedInstance];
  
  [server requireRegions:^(NSError *error) {
    if (error) {
      DLog(@"Error filling in stop: %@", error.localizedDescription);
      completion(error);
      return;
    }
    
    SVKRegion *region = stop.region;
    if (! region) {
      // We have regions, but this stop doesn't match any known region
      completion([TKBuzzInfoProvider errorForUserForBrokenStop]);
      return;
    }
    
    // construct the parameters
    NSMutableDictionary *paras = [NSMutableDictionary dictionary];
    paras[@"region"] = stop.regionName;
    paras[@"code"] = stop.stopCode;
    paras[@"name"] = stop.name ?: name;
    paras[@"modeInfo"] = stop.stopModeInfo;
    
    if (stop.location) {
      CLLocationCoordinate2D coordinate = stop.coordinate;
      if (CLLocationCoordinate2DIsValid(coordinate)) {
        paras[@"lat"] = @(coordinate.latitude);
        paras[@"lng"] = @(coordinate.longitude);
      }
    }
    
    [server hitSkedGoWithMethod:@"POST"
                           path:@"stopFinder.json"
                     parameters:paras
                         region:region
                 callbackOnMain:NO
                        success:
     ^(NSInteger status, id responseObject, NSData *data) {
#pragma unused(status, data)
        // set the stop properties
       [stop.managedObjectContext performBlock:^{
         BOOL success = [TKBuzzInfoProvider addStop:stop fromResponse:responseObject];
         ZAssert(success, @"Error processing: %@", responseObject);
         completion(nil);
       }];
     }
                               failure:
     ^(NSError *anotherError) {
       dispatch_async(dispatch_get_main_queue(), ^{
         DLog(@"Error response: %@", anotherError);
         completion(anotherError);
       });
     }];
  }];
}

- (void)addContentToService:(Service *)service
               fromResponse:(NSDictionary *)responseDict
{
  NSParameterAssert(service);
  NSParameterAssert(responseDict);
  
  ZAssert(service.managedObjectContext.parentContext != nil || [NSThread isMainThread], @"Not on the right thread!");
  
  NSManagedObjectContext *context = service.managedObjectContext;
  
  // real time status
  NSString *realTimeStatus = responseDict[@"realTimeStatus"];
  if (realTimeStatus) {
    [TKParserHelper adjustService:service
          forRealTimeStatusString:realTimeStatus];
  }
  
  // real time vehicles
  [TKAPIToCoreDataConverter updateVehiclesForService:service
                                      primaryVehicle:responseDict[@"realtimeVehicle"]
                                 alternativeVehicles:responseDict[@"realtimeVehicleAlternatives"]];
  
  // alert
  [TKAPIToCoreDataConverter updateOrAddAlerts:responseDict[@"alerts"]
                             inTripKitContext:context];
  
  // mode info
  ModeInfo *modeInfo = [ModeInfo modeInfoForDictionary:responseDict[@"modeInfo"]];
  
  // accessibility
  if ([responseDict[@"wheelchairAccessible"] boolValue]) {
    service.wheelchairAccessible = true;
  }
  if ([responseDict[@"bicycleAccessible"] boolValue]) {
    service.bicycleAccessible = true;
  }
  
  // parse the shapes
  NSArray *shapesArray = responseDict[@"shapes"];
  [TKParserHelper insertNewShapes:shapesArray
                       forService:service
                     withModeInfo:modeInfo];
}

#pragma mark - Private methods

+ (BOOL)addStop:(StopLocation *)stop
   fromResponse:(id)responseObject
{
  if ([responseObject count] == 0) {
    return NO;
  }
  ZAssert(stop.managedObjectContext.parentContext != nil || [NSThread isMainThread], @"Not on the right thread!");
  
  NSManagedObjectContext *tripKitContext = stop.managedObjectContext;
  
  NSArray *groups = responseObject[@"groups"];
  if (groups) {
    for (NSDictionary *groupDict in groups) {
      NSString *key = groupDict[@"key"];
      if (! [key isEqualToString:stop.stopCode])
        continue;
      
      NSArray *stopList = groupDict[@"stops"];
      for (NSDictionary *stopDict in stopList) {
        NSString *code = stopDict[@"code"];
        
        // is this our stop?
        if ([stop.stopCode isEqualToString:code]) {
          [TKAPIToCoreDataConverter updateStopLocation:stop
                                        fromDictionary:stopDict];
          
        } else {
          // we always add all the stops, because the cell is new
          StopLocation *newStop = [TKAPIToCoreDataConverter insertNewStopLocation:stopDict
                                                                 inTripKitContext:tripKitContext];
          
          // make sure we have an ID
          NSError *error = nil;
          [tripKitContext obtainPermanentIDsForObjects:@[newStop] error:&error];
          ZAssert(! error, @"Error obtaining permanent ID for '%@': %@", newStop, error);
        }
      }
      return stopList.count > 0;
    }
    return NO;
  
  } else {
    [TKAPIToCoreDataConverter updateStopLocation:stop fromDictionary:responseObject];
    return YES;
  }
}

@end

File: Classes/server/TKBuzzInfoProvider.swift
//
//  TKBuzzInfoProvider.swift
//  TripGo
//
//  Created by Adrian Schoenig on 11/12/2015.
//  Copyright © 2015 SkedGo Pty Ltd. All rights reserved.
//

import Foundation

import Marshal
import RxSwift


// MARK: - Fetcher methods -

extension TKBuzzInfoProvider {
  
  /**
   Asynchronously fetches additional region information for the provided region.
   
   - Note: Completion block is executed on the main thread.
   */
  public class func fetchRegionInformation(forRegion region: SVKRegion, completion: @escaping (TKRegionInfo?) -> Void)
  {
    SVKServer.fetchArray(TKRegionInfo.self,
                         method: .POST, path: "regionInfo.json",
                         parameters: ["region": region.name],
                         region: region,
                         keyPath: "regions")
    { regions in
      completion(regions.first)
    }
  }
  
  /**
   Asynchronously fetches paratransit information for the provided region.
   
   - Note: Completion block is executed on the main thread.
   */
  public class func fetchParatransitInformation(forRegion region: SVKRegion, completion: @escaping (TKParatransitInfo?) -> Void)
  {
    fetchRegionInformation(forRegion: region) { info in
      completion(info?.paratransitInformation)
    }
  }
  
  /**
   Asynchronously fetches all available individual public transport modes for the provided region.
   
   - Note: Completion block is executed on the main thread.
   */
  public class func fetchPublicTransportModes(forRegion region: SVKRegion, completion: @escaping ([ModeInfo]) -> Void)
  {
    fetchRegionInformation(forRegion: region) { info in
      completion(info?.transitModes ?? [])
    }
  }
  
  /**
   Asynchronously fetches additional location information for a specified coordinate.
   
   - Note: Completion block is executed on the main thread.
   */
  public class func fetchLocationInformation(_ coordinate: CLLocationCoordinate2D, forRegion region: SVKRegion, completion: @escaping (TKLocationInfo?) -> Void) {
    
    let paras: [String: Any] = [
      "lat": coordinate.latitude,
      "lng": coordinate.longitude
    ]
    
    SVKServer.fetch(TKLocationInfo.self,
                    path: "locationInfo.json",
                    parameters: paras,
                    region: region,
                    completion: completion)
  }
  
  /**
   Asynchronously fetches transit alerts for the provided region.
   
   - Note: Completion block is executed on the main thread.
   */
  public class func fetchTransitAlerts(forRegion region: SVKRegion, completion: @escaping ([TKSimpleAlert]) -> Void) {
    
    SVKServer.fetchArray(TKSimpleAlert.self,
                         path: "alerts/transit.json",
                         parameters: ["region": region.name],
                         region: region,
                         keyPath: "alerts",
                         completion: completion)
  }
  
  /**
   Asynchronously fetches transit alerts for the provided region using Rx.
   */
  public class func rx_fetchTransitAlerts(forRegion region: SVKRegion) -> Observable<[TKAlert]> {
    let paras: [String: Any] = [
      "region": region.name as Any
    ]
    
    return SVKServer.sharedInstance().rx
      .hit(.GET, path: "alerts/transit.json", parameters: paras, region: region)
      .map { (_, response) -> [TKAlert] in
        if let json = response as? [String: Any] {
          let alerts: [TKSimpleAlert]? = try? json.value(for: "alerts")
          return alerts ?? []
        } else {
          return []
        }
    }
  }
}


// MARK: - Helper Extensions -

extension SVKServer {
  
  fileprivate class func fetch<E: Unmarshaling>(
    _ type: E.Type,
    method: HTTPMethod = .GET,
    path: String,
    parameters: [String: Any]? = nil,
    region: SVKRegion,
    keyPath: String? = nil,
    completion: @escaping (E?) -> Void
  )
  {
    SVKServer.sharedInstance().hitSkedGo(
      withMethod: method.rawValue,
      path: path,
      parameters: parameters,
      region: region,
      success: { _, response in
        guard let json = response as? [String: Any] else {
          preconditionFailure() // FIXME
        }
        do {
          let result: E
          if let keyPath = keyPath {
            result = try json.value(for: keyPath)
          } else {
            result = try E(object: json)
          }
          completion(result)
        } catch {
          SGKLog.debug("TKBuzzInfoProvider") { "Encountered \(error), when fetching \(path), paras: \(parameters)" }
          completion(nil)
        }
    },
      failure: { error in
        SGKLog.debug("TKBuzzInfoProvider") { "Encountered \(error), when fetching \(path), paras: \(parameters)" }
        completion(nil)
    })
  }
  
  fileprivate class func fetchArray<E: Unmarshaling>(
    _ type: E.Type,
    method: HTTPMethod = .GET,
    path: String,
    parameters: [String: Any]? = nil,
    region: SVKRegion,
    keyPath: String? = nil,
    completion: @escaping ([E]) -> Void
    )
  {
    SVKServer.sharedInstance().hitSkedGo(
      withMethod: method.rawValue,
      path: path,
      parameters: parameters,
      region: region,
      success: { _, response in
        guard let json = response as? [String: Any] else {
          preconditionFailure() // FIXME
        }
        do {
          let result: [E]
          if let keyPath = keyPath {
            result = try json.value(for: keyPath)
          } else {
            result = try Array<E>.value(from: json)
          }
          completion(result)
        } catch {
          SGKLog.debug("TKBuzzInfoProvider") { "Encountered \(error), when fetching \(path), paras: \(parameters)" }
          completion([])
        }
    },
      failure: { error in
        SGKLog.debug("TKBuzzInfoProvider") { "Encountered \(error), when fetching \(path), paras: \(parameters)" }
        completion([])
    })
  }
  
}
